src/catchup/simulation/ApplyTransactionsWork.cpp

```diff
@@ -23,7 +31,184 @@ ApplyTransactionsWork::ApplyTransactionsWork(
     , mResultHistory{}
     , mResultIter(mResultHistory.txResultSet.results.cend())
     , mMaxOperations(desiredOperations)
+    , mMultiplier(multiplier)
+    , mVerifyResults(verifyResults)
+    , mSimulateOnly(simulateOnly)
 {
+    if (mMultiplier == 0)
+    {
+        throw std::runtime_error("Invalid multiplier!");
+    }
+}
+
+static void
+checkResults(Application& app, uint32_t ledger,
+             std::vector<TransactionResultPair> const& results)
+{
+    auto resSet = TransactionFrame::getTransactionHistoryResults(
+        app.getDatabase(), ledger);
+
+    assert(resSet.results.size() == results.size());
+    for (size_t i = 0; i < results.size(); i++)
+    {
+        auto const& res = results[i];
+        assert(res.transactionHash == resSet.results[i].transactionHash);
+
+        auto const& dbRes = resSet.results[i].result;
+        if (dbRes.result.code() != res.result.result.code())
+        {
+            auto msg = fmt::format(
+                "Expected result code {} does not agree with {} for tx {}",
+                res.result.result.code(), dbRes.result.code(),
+                binToHex(res.transactionHash));
+            CLOG(ERROR, "History") << msg;
+        }
+    }
+}
+
+bool
+ApplyTransactionsWork::hasSig(PublicKey const& account,
+                              TransactionEnvelope const& oldEnvelope,
+                              Hash const& hash)
+{
+    // Is the signature of this account present in the envelope we're
+    // simulating?
+    if (!mUsedPubKeys.insert(account).second)
+    {
+        return false;
+    }
+
+    for (auto const& sig : oldEnvelope.signatures)
+    {
+        if (PubKeyUtils::verifySig(account, sig.signature, hash))
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+uint32_t
+ApplyTransactionsWork::scaleLedger(
+    std::vector<TransactionEnvelope>& transactions,
+    std::vector<TransactionResultPair>& results,
+    std::vector<UpgradeType>& upgrades)
+{
+    assert(mTransactionIter != mTransactionHistory.txSet.txs.cend());
+    assert(mResultIter != mResultHistory.txResultSet.results.cend());
+
+    LedgerTxn ltx(mApp.getLedgerTxnRoot());
+
+    // No amplification needed, simply return existing transactions and results
+    if (mCount == 0)
```

Comment:
`mCount` should be an argument here and a local variable at the call site, rather than an instance variable. It will make it easier to reason about because then maintainers won't be searching through the file for other uses like I just was.